tests/net/lib/lwm2m/engine/src/lwm2m_observation.c

```diff
@@ -0,0 +1,532 @@
+{
+	struct lwm2m_obj_path_list *entry, *tmp;
+
+	if (name != NULL) {
+		TEST_VERBOSE_PRINT("Path List %s:\n", name);
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(lwm2m_path_list, entry, tmp, node) {
+		char buf[LWM2M_MAX_PATH_STR_LEN];
+
+		lwm2m_path_to_string(buf, LWM2M_MAX_PATH_STR_LEN, &entry->path, entry->path.level);
+		TEST_VERBOSE_PRINT("- %s\n", buf);
+	}
+}
+
+static void run_insertion_test(char const *insert_path_str[], int insertions_count,
+				   char const *expected_path_str[])
+{
+	/* GIVEN: different paths */
+	struct lwm2m_obj_path_list lwm2m_path_list_buf[insertions_count];
+	sys_slist_t lwm2m_path_list;
+	sys_slist_t lwm2m_path_free_list;
+	int ret;
+
+	lwm2m_engine_path_list_init(&lwm2m_path_list, &lwm2m_path_free_list, lwm2m_path_list_buf,
+					insertions_count);
+
+	/* WHEN: inserting each path */
+	struct lwm2m_obj_path insert_path;
+	char name[20];
+
+	for (int i = 0; i < insertions_count; ++i) {
+		ret = lwm2m_string_to_path(insert_path_str[i], &insert_path, '/');
+		zassert_true(ret >= 0, "Conversion to path #%d failed", i);
+
+		ret = lwm2m_engine_add_path_to_list(&lwm2m_path_list, &lwm2m_path_free_list,
+							&insert_path);
+
+		sprintf(name, "Insertion: %d", i);
+		print_path_list(&lwm2m_path_list, name);
+		zassert_true(ret >= 0, "Insertion #%d failed", i);
+		/* THEN: path order is maintained */
+		assert_path_list_order(&lwm2m_path_list);
+	}
+
+	print_path_list(&lwm2m_path_list, "Final");
+
+	/* AND: final list matches expectation */
+	struct lwm2m_obj_path_list *entry, *tmp;
+	int path_num = 0;
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&lwm2m_path_list, entry, tmp, node) {
+		struct lwm2m_obj_path expected_path;
+
+		lwm2m_string_to_path(expected_path_str[path_num++], &expected_path, '/');
+		zassert_mem_equal(&entry->path, &expected_path, sizeof(struct lwm2m_obj_path),
+				  "Path #%d did not match expectation", path_num);
+	}
+}
+
+ZTEST(lwm2m_observation, test_add_path_to_list)
+{
+	/* clang-format off */
+	char const *insert_path_str[] = {
+		LWM2M_PATH(2),
+		LWM2M_PATH(1),
+		LWM2M_PATH(1, 2),
+		LWM2M_PATH(1, 1),
+		LWM2M_PATH(1, 1, 10),
+		LWM2M_PATH(1, 1, 10, 10),
+		LWM2M_PATH(1, 1, 10, 9),
+		LWM2M_PATH(1, 2, 10, 11),
+
+		LWM2M_PATH(100),
+
+		LWM2M_PATH(41),
+		LWM2M_PATH(43, 3),
+		LWM2M_PATH(45, 2, 2),
+		LWM2M_PATH(47, 1, 1, 1),
+
+		LWM2M_PATH(57, 1, 1, 1),
+		LWM2M_PATH(55, 2, 2),
+		LWM2M_PATH(53, 3),
+		LWM2M_PATH(51),
+	};
+
+	char const *expected_path_str[] = {
+		LWM2M_PATH(1),
+		LWM2M_PATH(1, 1),
+		LWM2M_PATH(1, 1, 10),
+		LWM2M_PATH(1, 1, 10, 9),
+		LWM2M_PATH(1, 1, 10, 10),
+		LWM2M_PATH(1, 2),
+		LWM2M_PATH(1, 2, 10, 11),
+		LWM2M_PATH(2),
+		LWM2M_PATH(41),
+		LWM2M_PATH(43, 3),
+		LWM2M_PATH(45, 2, 2),
+		LWM2M_PATH(47, 1, 1, 1),
+		LWM2M_PATH(51),
+		LWM2M_PATH(53, 3),
+		LWM2M_PATH(55, 2, 2),
+		LWM2M_PATH(57, 1, 1, 1),
+		LWM2M_PATH(100),
+	};
+	/* clang-format on */
+
+	run_insertion_test(insert_path_str, 21, expected_path_str);
```

Comment:
The actual test entry count is 17, this causes seg fault with clang (see CI). Perhaps it'd be safer to use `ARRAY_SIZE(insert_path_str)` to avoid such mistakes.